### PR TITLE
metrics code optimization

### PR DESCRIPTION
**Pull request by wxbty**
## What is the purpose of the change

When I was writing registry metric implement, I will  use some codes of metrics-api, and found that some codes might be optimized.

## Brief changelog

Mainly optimize the following aspects:

1. Object-oriented design optimization

- The quadruple of interfaceName, methodName, group, version, which appears frequently and  use together, can be replaced by an object
- Both DefaultMetricsCollector and MetricsStatComposite have the behavior of publishing MetricsEvent. MetricsStatComposite locates the data aggregation carrier and should only be responsible for the processing of internal data. Event registration and publishing should be managed by a separate class (refer to spring), and the separation of class responsibilities can be easier Scale and prevent single-class code bloat.

2. Optimization of memory usage

- Every time the method (invoke, onResponse, onError) in MetricsFilter is triggered, the new temporary object MetricsCollectExecutor will be created. This is unnecessary. Changing to the statc method can reduce the number of yong gc.

3. Optimization of some repetitive codes

主要对以下几个方面做了优化：

1、面向对象的设计优化

- interfaceName,  methodName,  group,  version 这个四元组，出现频率高且共同使用，可以使用对象代替
- DefaultMetricsCollector 和 MetricsStatComposite 两个类都有发布MetricsEvent的行为，MetricsStatComposite定位数据聚合载体，应该只负责内部数据的处理，事件注册和发布应该由单独的类来管理（参考spring），类职责分离可以更容易扩展和防止单类代码膨胀。

2、内存使用的优化

-    MetricsFilter里的方法（invoke,onResponse,onError）每次触发，会new临时对象MetricsCollectExecutor，这是没必要的，改成statc方法可以减少yong gc次数。

3、其他
一些重复代码的优化；
定义空事件类型，从而发布事件时不用考虑空指针异常；

### CODE REVIEW

**Comment by jojocodeX** [comment]
> ## What is the purpose of the change
> When I was writing registry metric implement, I will use some codes of metrics-api, and found that some codes might be optimized.
> 
> ## Brief changelog
> Mainly optimize the following aspects:
> 
> 1. Object-oriented design optimization
> 
> * The quadruple of interfaceName, methodName, group, version, which appears frequently and  use together, can be replaced by an object
> * Both DefaultMetricsCollector and MetricsStatComposite have the behavior of publishing MetricsEvent. MetricsStatComposite locates the data aggregation carrier and should only be responsible for the processing of internal data. Event registration and publishing should be managed by a separate class (refer to spring), and the separation of class responsibilities can be easier Scale and prevent single-class code bloat.
> 
> 2. Optimization of memory usage
> 
> * Every time the method (invoke, onResponse, onError) in MetricsFilter is triggered, the new temporary object MetricsCollectExecutor will be created. This is unnecessary. Changing to the statc method can reduce the number of yong gc.
> * In addition, a simple object pool is used to obtain the previous quadruple object
> 
> 3. Optimization of some repetitive codes
> 
> 主要对以下几个方面做了优化：
> 
> 1、面向对象的设计优化
> 
> * interfaceName,  methodName,  group,  version 这个四元组，出现频率高且共同使用，可以使用对象代替
> * DefaultMetricsCollector 和 MetricsStatComposite 两个类都有发布MetricsEvent的行为，MetricsStatComposite定位数据聚合载体，应该只负责内部数据的处理，事件注册和发布应该由单独的类来管理（参考spring），类职责分离可以更容易扩展和防止单类代码膨胀。
> 
> 2、内存使用的优化
> 
> * MetricsFilter里的方法（invoke,onResponse,onError）每次触发，会new临时对象MetricsCollectExecutor，这是没必要的，改成statc方法可以减少yong gc次数。
> * 另外，使用了简单的对象池，来获取前面的四元组对象
> 
> 3、其他 一些重复代码的优化； 定义空事件类型，从而发布事件时不用考虑空指针异常；

这个优化是有必要的，比如consumer这侧的指标采集，consumer这侧和provider这侧其实有很多相同的地方，现在这种方式使用上不是太友好。对于不同场景下指标收集，api使用上不方便

File: dubbo-dependencies-bom/pom.xml [hunk 1]
```diff
@@ -184,7 +184,7 @@
         <portlet_version>2.0</portlet_version>
         <maven_flatten_version>1.1.0</maven_flatten_version>
         <commons_compress_version>1.21</commons_compress_version>
-        <revision>3.2.0-beta.4-SNAPSHOT</revision>
+        <revision>3.2.0-beta.11-SNAPSHOT</revision>
```
Comment (AlbumenJ) [inline dubbo-dependencies-bom/pom.xml:187]: revert this line

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-dependencies-bom/pom.xml (hunk 1 above)
Comment (wxbty) [reply dubbo-dependencies-bom/pom.xml]: done

File: dubbo-common/src/main/java/org/apache/dubbo/common/metrics/model/ServiceInfo.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    private String version;
+
+    private ServiceInfo(String interfaceName, String methodName, String group, String version) {
+        this.interfaceName = interfaceName;
+        this.methodName = methodName;
+        this.group = group;
+        this.version = version;
+    }
+
+    public static ServiceInfo getInstance(String interfaceName, String methodName, String group, String version) {
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/metrics/model/ServiceInfo.java:40]: Can we just simply use `Invocation` here?

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/metrics/model/ServiceInfo.java (hunk 2 above)
Comment (wxbty) [reply dubbo-common/src/main/java/org/apache/dubbo/common/metrics/model/ServiceInfo.java:40]: Looking at the class description that Invocation is non-thread-safe, I'm not sure if the property will be modified by another unknown thread during use. Another developer, seeing that MethodMetric relies on a Invocation, might not immediately guess what the relationship is?

**Comment by wxbty** [comment]
@AlbumenJ PTAL

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Comment by AlbumenJ** [comment]
@songxiaosheng PTAL

**Review by songxiaosheng — APPROVED** [review]
LGTM

File: dubbo-common/src/main/java/org/apache/dubbo/common/metrics/event/MetricsEventMulticaster.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.apache.dubbo.common.metrics.event;
+
+import org.apache.dubbo.common.metrics.listener.MetricsListener;
```
[the remark this answers is not in the document]
Comment (wxbty) [reply dubbo-common/src/main/java/org/apache/dubbo/common/metrics/event/MetricsEventMulticaster.java]: Please see the last submission, the same as the following

**Review by songxiaosheng — APPROVED** [review]
LGTM

**Pull request merged**